File: scripts/update_just_skill.py
# /// script
# dependencies = [
#   "requests",
#   "markdown-it-py",
#   "mdformat",
# ]
# ///

import os
import requests
from markdown_it import MarkdownIt
from pathlib import Path

README_URL = 'https://raw.githubusercontent.com/casey/just/refs/heads/master/README.md'
# Adjusting path to match your logic: ../skills/justfile/SKILL.md
SKILL_FILE = Path(__file__).parent.parent / "skills" / "justfile" / "SKILL.md"

EXCLUDE_SECTIONS = [
    'installation', 'backwards compatibility', 'editor support', 
    'changelog', 'miscellanea', 'contributing', 
    'frequently asked questions', 'further ramblings', 'packages'
]

INCLUDE_SECTIONS = [
    'quick start', 'examples', 'features', 'the default recipe'
]

def filter_markdown(content):
    md = MarkdownIt("gfm-like")
    tokens = md.parse(content)
    
    new_tokens = []
    keeping = True
    
    # Iterate through tokens to filter sections
    i = 0
    while i < len(tokens):
        token = tokens[i]
        
        # Detect Headings
        if token.type == "heading_open":
            # The actual text is usually in the next 'inline' token
            inline_token = tokens[i+1]
            text = inline_token.content.lower().strip()
            
            if any(text == s or text.startswith(s + ' ') for s in EXCLUDE_SECTIONS):
                keeping = False
            elif any(text == s or text.startswith(s + ' ') for s in INCLUDE_SECTIONS):
                keeping = True
        
        # Filter HTML (Table of Contents / Shields)
        if token.type == "html_block":
            if "Table of Contents" in token.content or "img.shields.io" in token.content:
                i += 1
                continue

        if keeping:
            new_tokens.append(token)
        i += 1
        
    # Simple reconstruction (for a more advanced rebuild, use mdformat)
    # Since we are filtering blocks, we can just extract the source map or use a renderer
    return md.renderer.render(new_tokens, md.options, {})

def main():
    print("Fetching README...")
    resp = requests.get(README_URL)
    resp.raise_for_status()
    markdown_text = resp.text

    print("Processing...")
    # Note: markdown-it-py renders to HTML by default. 
    # For a direct MD-to-MD filter like your JS script, we use the logic below:
    lines = markdown_text.splitlines()
    filtered_lines = []
    keeping = True
    
    # Simpler line-based approach for Markdown-to-Markdown preservation
    for line in lines:
        if line.startswith('#'):
            title = line.lstrip('#').strip().lower()
            if any(title == s or title.startswith(s + ' ') for s in EXCLUDE_SECTIONS):
                keeping = False
            elif any(title == s or title.startswith(s + ' ') for s in INCLUDE_SECTIONS):
                keeping = True
        
        if keeping:
            # Skip TOC and Shields
            if "Table of Contents" in line or "img.shields.io" in line:
                continue
            filtered_lines.append(line)

    new_content = "\n".join(filtered_lines)

    print(f"Reading {SKILL_FILE}...")
    skill_path = Path(SKILL_FILE)
    skill_text = skill_path.read_text(encoding="utf-8")

    MARKER = 'It fully documents the Justfile syntax and system.\n\n---'
    if MARKER in skill_text:
        header = skill_text.split(MARKER)[0] + MARKER
    else:
        print("Marker not found! Appending.")
        header = skill_text + ("\n\n---" if not skill_text.strip().endswith("---") else "")

    final_output = f"{header}\n\n{new_content}"
    
    print("Writing...")
    skill_path.write_text(final_output, encoding="utf-8")
    print("Done.")

if __name__ == "__main__":
    main()

File: scripts/update_just_skill.ts
#!/usr/bin/env bun

import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import remarkGfm from 'remark-gfm';
import fs from 'node:fs';
import path from 'node:path';

const README_URL = 'https://raw.githubusercontent.com/casey/just/refs/heads/master/README.md';
const SKILL_FILE = path.join(import.meta.dir, '..', 'skills', 'justfile', 'SKILL.md');

// Sections to explicitly exclude
const EXCLUDE_SECTIONS = [
  'installation',
  'backwards compatibility',
  'editor support',
  'changelog',
  'miscellanea',
  'contributing',
  'frequently asked questions',
  'further ramblings',
  'packages',
];

// Sections to explicitly include
const INCLUDE_SECTIONS = [
  'quick start',
  'examples',
  'features',
  'the default recipe',
];

function filterSections() {
  return (tree) => {
    const newChildren = [];
    let keeping = true;

    for (const node of tree.children) {
      if (node.type === 'heading') {
        const text = node.children.map(c => c.value).join('').toLowerCase().trim();
        
        if (EXCLUDE_SECTIONS.some(s => text === s || text.startsWith(s + ' '))) {
          keeping = false;
        } else if (INCLUDE_SECTIONS.some(s => text === s || text.startsWith(s + ' '))) {
          keeping = true;
        }
      }

      if (keeping && node.type === 'html') {
        if (node.value.includes('Table of Contents') || 
            (node.value.includes('align=center') && node.value.includes('img.shields.io'))) {
          continue;
        }
      }

      if (keeping) {
        newChildren.push(node);
      }
    }
    tree.children = newChildren;
  };
}

async function main() {
  console.log('Fetching README...');
  const response = await fetch(README_URL);
  if (!response.ok) throw new Error(`Failed to fetch: ${response.statusText}`);
  const markdown = await response.text();

  console.log('Processing Markdown...');
  const processor = unified()
    .use(remarkParse)
    .use(remarkGfm)
    .use(filterSections)
    .use(remarkStringify);

  const processed = await processor.process(markdown);
  const newContent = processed.toString();

  console.log('Reading SKILL.md...');
  let skillContent = fs.readFileSync(SKILL_FILE, 'utf-8');

  const MARKER = 'It fully documents the Justfile syntax and system.\n\n---';
  const markerIndex = skillContent.indexOf(MARKER);

  let header;
  if (markerIndex !== -1) {
    header = skillContent.substring(0, markerIndex + MARKER.length);
  } else {
    console.warn('Marker not found! Appending to end.');
    header = skillContent;
    if (!header.trim().endsWith('---')) {
        header += '\n\n---';
    }
  }

  const finalContent = header + '\n\n' + newContent;

  console.log('Writing to SKILL.md...');
  fs.writeFileSync(SKILL_FILE, finalContent);
  console.log('Done.');
}

main().catch(console.error);
